worldengine/simulations/precipitation.py

```diff
@@ -42,18 +47,61 @@ def _calculate(seed, width, height):
 
         for y in range(height):#TODO: numpy
             y_scaled = float(y) / height
+            n_scale = 1024 / float(height)
             latitude_factor = 1.0 - (abs(y_scaled - 0.5) * 2)
             for x in range(width):
-                n = snoise2(x / freq, y / freq, octaves, base=base)
+                n = snoise2((x * n_scale) / freq, (y * n_scale) / freq, octaves, base=base)
 
                 # Added to allow noise pattern to wrap around right and left.
                 if x < border:
-                    n = (snoise2(x / freq, y / freq, octaves,
+                    n = (snoise2( (x * n_scale) / freq, (y * n_scale) / freq, octaves,
                                  base=base) * x / border) + (
-                        snoise2((x + width) / freq, y / freq, octaves,
+                        snoise2(( (x * n_scale) + width) / freq, (y * n_scale) / freq, octaves,
                                 base=base) * (border - x) / border)
 
-                precipitation = (latitude_factor + n * 4) / 5.0
-                precipitations[y, x] = precipitation
+                precipitations[y, x] = n
+        min_precip = None
+        max_precip = None
+        min_temp = None
+        max_temp = None
 
+        #find ranges
+        for y in range(height):#TODO: numpy
+            for x in range(width):
+                t = world.temperature_at((x, y))
+                p = precipitations[y, x]
+                if min_precip is None or p < min_precip:
+                    min_precip = p
+                if max_precip is None or p > max_precip:
+                    max_precip = p
+                if min_temp is None or t < min_temp:
+                    min_temp = t
+                if max_temp is None or t > max_temp:
+                    max_temp = t
+        precip_delta = (max_precip - min_precip)
+        temp_delta = (max_temp - min_temp)
+
+        #modify precipitation based on temperature
+        for y in range(height):
+            for x in range(width):
+                t = (world.temperature_at((x, y)) - min_temp) / temp_delta
+                p = (precipitations[y, x] - min_precip) / precip_delta
+                precipitations[y, x] = (2 * p * (math.pow((t),curve_gamma) + curve_bonus))
+
+        #Renormalize temperatures because the temperature 
+        #changes will probably not fully extend from -1 to 1.
+        min_temp = None
+        max_temp = None
+        for y in range(height):
+            for x in range(width):
+                p = precipitations[y, x]
+                if min_precip is None or p < min_precip:
+                    min_precip = p
+                if max_precip is None or p > max_precip:
+                    max_precip = p
+        precip_delta = (max_precip - min_precip)
+        for y in range(height):
+            for x in range(width):
+                precipitations[y, x] = (2 * ((precipitations[y, x] - min_precip) / precip_delta)) - 1
```

Comment:
I think that everything from line 63 to here can be replaced by numpy-code. Rule of thumb: Whenever you are iterating over an array, numpy can probably do it 100 times faster.
I would highly prefer if you could change the code accordingly (I can provide an elaborate example, if you need it). It was quite some work getting all the numpy arrays into place and getting rid of most of the iterations etc. throughout all of the code, I would hate to start the fight all over.